<a href="https://colab.research.google.com/github/superoverflow/notebooks/blob/main/DND_D20_Advantage_Disadvantage.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[DnD rules](https://bg3.wiki/wiki/Advantage) says if you have advantage, then you roll 2 D20 and take the max if you have a disadvantage, then you roll 2 D20 and take the min

Calculate the expected roll for advantage and disadvantage

In [9]:
# imports
import numpy as np

D20 can be represented as a numpy array

In [10]:
D20 = np.arange(1, 21)
D20

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20])

If we roll two D20, with first dice rolled 5, the expected outcome for another dice would be:

In [11]:
np.maximum(5, D20)

array([ 5,  5,  5,  5,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20])

We can first list out all possibility and calculate the expected value for advantage situation

In [12]:
A = D20
for i in range(2, 21):
  A = np.vstack((A, np.maximum(i, D20)))

A

array([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 2,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 3,  3,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 4,  4,  4,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 5,  5,  5,  5,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 6,  6,  6,  6,  6,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 7,  7,  7,  7,  7,  7,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 8,  8,  8,  8,  8,  8,  8,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [ 9,  9,  9,  9,  9,  9,  9,  9,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20],
       [11, 11, 11, 11, 11, 11

In [13]:
np.average(A)

13.825

Similiarly for disadvantage

In [14]:
D = np.full((1, 20), 1)
for i in range(2, 21):
  D = np.vstack((D, np.minimum(i, D20)))

D

array([[ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
         1,  1,  1,  1],
       [ 1,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
         2,  2,  2,  2],
       [ 1,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
         3,  3,  3,  3],
       [ 1,  2,  3,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,
         4,  4,  4,  4],
       [ 1,  2,  3,  4,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,
         5,  5,  5,  5],
       [ 1,  2,  3,  4,  5,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,
         6,  6,  6,  6],
       [ 1,  2,  3,  4,  5,  6,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,
         7,  7,  7,  7],
       [ 1,  2,  3,  4,  5,  6,  7,  8,  8,  8,  8,  8,  8,  8,  8,  8,
         8,  8,  8,  8],
       [ 1,  2,  3,  4,  5,  6,  7,  8,  9,  9,  9,  9,  9,  9,  9,  9,
         9,  9,  9,  9],
       [ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 10, 10, 10, 10, 10, 10,
        10, 10, 10, 10],
       [ 1,  2,  3,  4,  5,  6

In [15]:
np.average(D)

7.175

**Conclusion**

- For D20, the expected roll is 10.5
- For D20 advantage, the expected roll would be 13.825
- For D20 disadvantage, the expected roll would be 7.175


Compare with [Bless spell](https://bg3.wiki/wiki/Bless), which gives 1D4 and so the expected roll became 13. Advantage situation still do bit better